pysrc/bytewax/inputs.py

```diff
@@ -0,0 +1,204 @@
+"""Helpers to let you quickly define epoch / batching semantics.
+
+Use these to wrap an existing iterator which yields items.
+"""
+import datetime
+import heapq
+from operator import itemgetter
+from typing import Any, Iterable, Tuple, Callable
+
+
+def single_batch(wrap_iter: Iterable) -> Iterable[Tuple[int, Any]]:
+    """All input items are part of the same epoch.
+
+    Use this for non-streaming-style batch processing.
+
+    >>> list(single_batch(["a", "b", "c"]))
+    [(0, 'a'), (0, 'b'), (0, 'c')]
+
+    Args:
+
+        wrap_iter: Existing input iterable of just items.
+
+    Yields: Tuples of (epoch, item).
+    """
+    for item in wrap_iter:
+        yield (0, item)
+        
+
+def order(
+        wrap_iter: Iterable,
+        window_length: Any,
+        time_getter: Callable[[Any], Any] = itemgetter(0),
+        on_drop: Callable[[Any], None] = None,
+) -> Iterable[Tuple[int, Any]]:
+    """Order out-of-order values within a sliding window.
+
+    Values are emitted in order after seeing a new time that would
+    push them out of the sliding window. Values that arrive after they
+    would have been emitted are too late and are dropped.
+
+    Use if you have a possibly out-of-order data source as Bytewax
+    requires inputs to dataflows to be in epoch order.
+
+    >>> items = [
+    ...     {
+    ...         "timestamp": datetime.datetime(2022, 2, 22, 1, 2, 4),
+    ...         "value": "c",
+    ...     },
+    ...     {
+    ...         "timestamp": datetime.datetime(2022, 2, 22, 1, 2, 3),
+    ...         "value": "b",
+    ...     },
+    ...     {
+    ...         "timestamp": datetime.datetime(2022, 2, 22, 1, 2, 0),
+    ...         "value": "a",
+    ...     },
+    ... ]
+    >>> [
+    ...     item["value"]
+    ...     for item
+    ...     in order(
+    ...         items,
+    ...         datetime.timedelta(seconds=2),
+    ...         itemgetter("timestamp"),
+    ...     )
+    ... ]
+    ['b', 'c']
+
+    Args:
+    
+        wrap_iter: Existing input iterable.
+   
+        window_length: Maximum out-of-order-ness before dropping
```

Comment:
Trying to brainstorm the affirmative version of this to communicate "interval of opportunity".